### PR TITLE
Tests and support for eventfd syscall

### PR DESCRIPTION
**Pull request by robgjansen**
Add test cases and Shadow support for the `eventfd()` syscall and `read()` and `write()` on eventfd descriptors.

### CODE REVIEW

**Comment by robgjansen** [comment]
It looks like we are requiring that `rustfmt` is being run on new rust code, but I didn't find instructions in the docs about how to do it:
https://github.com/shadow/shadow/blob/dev/docs/coding-style.md

We should add the recommended method for formatting rust code.

(I ran `git clang-format` but didn't realize I needed to also run `rustfmt` separately.)

**Review by stevenengler — APPROVED** [review]
Looks good! Left a few minor comments.

File: src/main/host/syscall/eventfd.c [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+#ifdef DEBUG
+    /* This should always be a valid descriptor. */
+    int errcode = _syscallhandler_validateEventFDHelper(sys, efd, NULL);
+    if (errcode != 0) {
+        error("Unable to find eventfd %i that we just created.", efd);
+    }
+    utility_assert(errcode == 0);
+#endif
+
+    /* Set any options that were given. */
```
[the remark this answers is not in the document]
Comment (robgjansen) [reply src/main/host/syscall/eventfd.c:70]: Currently we store options in the `O_xxxxxxx` namespace in the "descriptor flags", and `EFD_SEMAPHORE` in the EFD namespace is incompatible (on my system, `EFD_SEMAPHORE`  has the same value as `O_WRONLY` (i.e., 1). I'm not sure which flags `fcntl` returns, but if it does return `EFD_SEMAPHORE` then I think we would need to overhaul how we store flags to be able to support this. I'm going to call this out of scope for this PR if you agree that's reasonable.

[the remark this answers is not in the document]
Comment (stevenengler) [reply src/main/host/syscall/eventfd.c:70]: Ah okay that makes sense, thanks. And I agree that it's out of scope.

File: src/test/eventfd/test_eventfd.rs [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+fn check_read_eagain(efd: RawFd) -> Result<(), String> {
+    let mut bytes: [u8; 8] = [0; 8];
+
+    test_utils::result_assert(
+        match read(efd, &mut bytes) {
+            Ok(_) => false,
+            Err(_) => true,
+        } && test_utils::get_errno() == libc::EAGAIN,
```
[the remark this answers is not in the document]
Comment (robgjansen) [reply src/test/eventfd/test_eventfd.rs:113]: Done, thank you!

File: src/test/eventfd/test_eventfd.rs [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+    let flags = [
+        EfdFlags::empty(),
+        EfdFlags::EFD_SEMAPHORE,
+        EfdFlags::EFD_NONBLOCK,
+        EfdFlags::EFD_CLOEXEC,
+        EfdFlags::EFD_SEMAPHORE | EfdFlags::EFD_NONBLOCK,
+    ];
+
+    for &flag in flags.iter() {
+        let efd: RawFd = eventfd(0, flag).unwrap();
```
[the remark this answers is not in the document]
Comment (robgjansen) [reply src/test/eventfd/test_eventfd.rs:70]: I'm going to stick with the String for now but can switch to Box when my rust is better (or we have a full example) :)

[the remark this answers is not in the document]
Comment (robgjansen) [reply src/test/eventfd/test_eventfd.rs:70]: Done, thank you!

File: src/main/host/syscall/eventfd.c [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+#include "support/logger/logger.h"
+
+///////////////////////////////////////////////////////////
+// Helpers
+///////////////////////////////////////////////////////////
+
+static int _syscallhandler_validateEventFDHelper(SysCallHandler* sys, int efd,
+                                                 EventD** event_desc_out) {
+    /* Check that fd is within bounds. */
+    if (efd <= 0) {
```
[the remark this answers is not in the document]
Comment (robgjansen) [reply src/main/host/syscall/eventfd.c:26]: Fixed, thanks!

File: src/test/eventfd/test_eventfd.rs [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+    check_read_eagain(efd)?;
+
+    // Writing u64_max-1 is allowed...
+    check_write_success(efd, u64::MAX - 1)?;
+    check_read_success(efd, u64::MAX - 1)?;
+    check_read_eagain(efd)?;
+    // ...but u64_max is not
+    check_write_einval(efd, u64::MAX)?;
+
+    close(efd).unwrap();
```
[the remark this answers is not in the document]
Comment (robgjansen) [reply src/test/eventfd/test_eventfd.rs]: Very useful function, thanks :)

**Pull request merged**